### ISSUE TITLE
Autoencoder + Classification Example in C#

### ISSUE THREAD

**Issue by whitmark** [Microsoft/CNTK#2879]
I was looking for an example (and some inspiration) to develop a general framework for a stacked autoencoder + classification solution in C# that would help reduce the ~100 numeric dimensions for a real-world binary classification problem and perhaps avoid the need to do feature scaling.  In researching this repository, I found similar requests below:

#3092
#672

It would seem this would be a good use-case for generic unsupervised multivariate problems and potentially a good candidate for a tutorial.  Instead of going through the trouble of synthetically generating sample data, one could use something like the [UCI Madelon Data Set](http://archive.ics.uci.edu/ml/datasets/madelon) which has 500 synthetic dimensions but only 20 are "real" and the data is already split into training, validation, and test sets.  Perhaps there is an alternative public data set where both dimensionality reduction and feature scaling are required.

1. Is it safe to assume the C# API can support this type of network to combine an autoecoder layer with a binary classification layer?
2. Does the use of autoencoding layers reduce/eliminate the need for feature scaling?
3. Would the CNTK team consider developing this type of training, validation, and evaluation example as a tutorial?

Thanks for considering.

**Comment by KeDengMS** [comment]
1. Yes, C# allows combination of different functions.
2. I think even with auto encoder, feature scaling/shifting is still preferred to make learning go faster per Geoffrey Hinton's [slides](http://www.cs.toronto.edu/~tijmen/csc321/slides/lecture_slides_lec6.pdf). Please try some experiments with or without prescaling.
3. More C# examples and better support for high level APIs are in our plan. Besides, CNTK is an open source effort, and we welcome any community contributes.

**Comment by whitmark** [comment]
Thanks @KeDengMS for the info and suggestion. My ~100 numeric input dimensions range from 10000 to 0.01 so it would seem that feature scaling/shifting could be useful, but I would like the model to evaluate individual sequences in realtime using a distributed C# app.  I assume if I do feature scaling, in addition to the model files, I would also need to distribute mean/stddev and/or min/max values for each input item calculated during training pre-processing to properly scale/shift inputs to be evaluated in realtime.  It was my hope that by using autoencoding layers the benefits of dimension reduction would far exceed those of feature scaling/shifting and I would only need to distribute the model files.  Perhaps there is a way to embed feature scaling/shifting data in the model for evaluation purposes that I am overlooking.

**Comment by whitmark** [comment]
Perhaps I should be more explicit.  Is there a reasonable way to embed into a saved model the statistics (e.g., min, max, mean, stddev) calculated for each input variable during pre-training so that this data can be extracted to apply feature scaling/shifting to the input series during the distributed evaluation process in C#?  In other words, can tensors be included in the saved model to just hold data for later extraction?

**Comment by KeDengMS** [comment]
Yes, batch normalization would do that if you don't know the data distribution up-front. Scale and bias would be computed at training time and stored into the model. Otherwise, you can have a built-in scale/shift for your data (like gray scale image from [0,255] to [0,1])

**Comment by whitmark** [comment]
Thanks @KeDengMS.  While reviewing Andrew Ng's lecture "[Batch Norm at Test Time](https://www.coursera.org/learn/deep-neural-network/lecture/FsoNw/batch-norm-at-test-time)", I see that batch normalization is tied to each mini-batch, so the mean and variance statistics need to be averaged across all mini-batches for the evaluation of an individual test sequence.  After reviewing the [BatchNormalization function docs](https://docs.microsoft.com/en-us/cognitive-toolkit/batchnormalization) and the specific C# function below, I see these running stats are maintained . . . very nice!

```C#
public static Function BatchNormalization(Variable operand,
                                            Variable scale,
                                            Variable bias,
                                            Variable runningMean,
                                            Variable runningInvStd,
                                            Variable runningCount,
                                            bool spatial,
                                            double normalizationTimeConstant = 0,
                                            double blendTimeConstant = 0,
                                            double epsilon = 0.00001,
                                            bool useCuDNNEngine = true,
                                            string name = "");

```
Batch norm might work fine, but in the event I would prefer to pre-calc and store my own population mean and variance stats in the model, how would I in C# go about inserting these data layers into the model before saving it and then extracting these layers after the model is loaded so I can properly scale/shift my ~100 numeric inputs (ranging from 10000 to 0.01) prior to evaluation?  Should they be built as hidden layers with no activation?

**Comment by KeDengMS** [comment]
After load your model, you can create a new graph by cloning the old one using ParameterCloningMethod.Share, and replace the inputs of your old graph with some computed output. This way you can extend your graph to have pre-calc.

**Comment by whitmark** [comment]
Thanks  @KeDengMS.  Based on another post you answered  (#2911) I found a [feed forward classifier end-to-end test using PerDimMeanVarianceNormalization in C#](https://github.com/Microsoft/CNTK/blob/master/Tests/EndToEndTests/CNTKv2CSharp/CNTKLibraryCSTrainingTest/SimpleFeedForwardClassifierTest.cs ).

It would seem that the PerDimMeanVarianceNormalization approach is better suited for normalizing input layers and storing these values per dim in the model vs BatchNormalization which is better suited for hidden layers applied after a connected layer.  Would you agree?

Since I don't see the PerDimMeanVarianceNormalization method in the [CNTK Library C# API listing](https://docs.microsoft.com/en-us/cognitive-toolkit/cntk-library-managed-api), will it continue to be supported going forward?

There are some useful approaches in these end-to-end tests that never make it do the docs or examples.

**Comment by KeDengMS** [comment]
IMO PerDimMeanVarianceNormalization falls in the category of preprocessing data, and this design follows BrainScript due to the limitation in what can be described. Once we are in Python, there are much more freedoms in preparing data, thus this feature is used less often.
In general, designing of API should balance the need of features and higher level abstraction. I think we should have some examples in Python showing user how to build a graph to compute input mean/variance, and then add the preprocessing to be part of graph. Adding new APIs that can be replaced by combination of existing ones is not ideal, because it would increase the maintenance cost as well as user's burden in memorizing the usage.

**Comment by whitmark** [comment]
Thanks @KeDengMS. It looks like the [feed forward classifier end-to-end test using PerDimMeanVarianceNormalization in C#](https://github.com/Microsoft/CNTK/blob/master/Tests/EndToEndTests/CNTKv2CSharp/CNTKLibraryCSTrainingTest/SimpleFeedForwardClassifierTest.cs ) replaces the inputs of the old graph with the normalized outputs albeit by a different approach than the ParameterCloningMethod.Share you suggested in an earlier post.  Perhaps this example can be refactored with preferred methods without too much effort.

Can you clarify whether BatchNormalization can be used for preprocessing inputs to retain the pre-calc in the graph?

**Comment by KeDengMS** [comment]
BatchNormalization(BN) is similar to PerDimMeanVarianceNormalization and you can give it a try. The difference is that BN normalizes input based on what has been fed, instead of the whole input data set. At training time, BN calculates stats for each minibatch, and blend it with running stats using normalization/blend time_constant.

**Comment by whitmark** [comment]
@KeDengMS , I see mean_variance_normalization has been added to python and onnx op.  Will this be added to the C# ops or is it duplicative of existing functionality.  (e.g., PerDimMeanVarianceNormalization)?

**Comment by KeDengMS** [comment]
You can find the implementation of these two functions [here](https://github.com/Microsoft/CNTK/blob/master/Source/CNTKv2LibraryDll/Function.cpp#L2370-L2402). mean_variance_normalization works on each minibatch individually.
